src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2337,11 +2332,36 @@ protected virtual MethodDefinition MarkMethod (MethodReference reference, Depend
 			if (Annotations.GetAction (method) == MethodAction.Nothing)
 				Annotations.SetAction (method, MethodAction.Parse);
 
+			ProcessRequiresUnreferencedCode (method, sourceLocationMember);
 			EnqueueMethod (method, reason);
 
 			return method;
 		}
 
+		protected internal void ProcessRequiresUnreferencedCode (MethodDefinition method, MessageOrigin sourceLocationMember)
+		{
+			// If the caller of a method is already marked with `RequiresUnreferencedCodeAttribute` a new warning should not
+			// be produced for the callee.
+			// Overrides of methods annotated with `RequiresUnreferencedCodeAttribute` should not produce a warning here,
+			// the warning should be put in the virtual method instead.
+			if (sourceLocationMember.MemberDefinition != null &&
+				Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (sourceLocationMember.MemberDefinition) ||
+				Annotations.GetBaseMethods (method) != null)
```

Comment:
I don't think I understand the last check about the base methods.
If the `method` is in a direct call (can be non-virtual in fact) then the warning should be reported.
I think the interesting question is what happens if I have:
```C#
interface ITest
{
    [RequiresUnreferencedCode]
    void DoDanger();
}

class Test : ITest
{
    [RequiredUnreferencedCode]
    void DoDanger() { }
}

void TestIt()
{
    ITest it = new Test();
    it.DoDanger(); // This should warn, but that is the easy part
}

// There should not be another warning about DoDanger even though there will be a MarkMethod(Test.DoDanger) probably coming from the interface as it's detected to be instantiated.
```

Reply:
> I don't think I understand the last check about the base methods.

It should prevent us from producing two warnings in cases such as:
```C#
Bar(typeof(DerivedClass));

void Bar([[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicMethods)] Type type) {}

class BaseClass
{
	[RequiresUnreferencedCode ("")]
	public virtual void Foo() {}
}

class DerivedClass : BaseClass
{
	[RequiresUnreferencedCode ("")]
	public override void Foo() {}
}
```

Reply:
I finally figured out the counter example:
```C#
		abstract class CovariantReturnBase
		{
			[RequiresUnreferencedCode ("Message for --CovariantReturnBase.GetRequiresUnreferencedCode--")]
			public abstract BaseReturnType GetRequiresUnreferencedCode ();
		}

		class CovariantReturnDerived : CovariantReturnBase
		{
			[RequiresUnreferencedCode ("Message for --CovariantReturnBase.GetRequiresUnreferencedCode--")]
			public override DerivedReturnType GetRequiresUnreferencedCode ()
			{
				return null;
			}
		}

		[LogDoesNotContain("--CovariantReturnBase.GetRequiresUnreferencedCode--")]
		[ExpectedWarning ("IL2026", "--CovariantReturnBase.GetRequiresUnreferencedCode--")]
		static void TestCovariantReturnCallOnDerived ()
		{
			var tmp = new CovariantReturnDerived ();
			tmp.GetRequiresUnreferencedCode ();
		}
```

The problem is that the callvirt in this case is not to the base method. The code as written in the MarkStep basically assumes that if an override is marked, it is never a target of a call - it's only because of virtual hierarchy marking. Which is mostly true for C#, since the compiler makes sure it generates callvirt to the base method regardless of which type the method was called on. This might not be true for non-C#, but it's definitely not true with covariant returns as shown above, in this case C# intentionally generates call to the override method.

The above test sample fails because there's no warning generated. Which is weird for multiple reasons, I'm still looking into it.

Reply:
Here is a change with the new test and modification to the markStep which makes all of the tests pass: https://github.com/mateoatr/linker/compare/requiresUnreferencedCode...vitek-karas:requiresUnreferencedCode